src/System.Text.Primitives/System/Text/InvariantParser_ulong.cs

```diff
@@ -30,20 +30,23 @@ public static bool TryParse(Utf8String utf8Text, out ulong value, out int bytesC
                         return true;
                     }
                 }
-                try
+                else if (value > UInt64.MaxValue / 10) // overflow
                 {
-                    ulong candidate = checked(value * 10 + nextByte - '0');
-                    Debug.Assert(candidate >= value);
-
-                    value = candidate;
-                    bytesConsumed++;
+                    value = 0;
+                    bytesConsumed = 0;
+                    return false;
                 }
-                catch (OverflowException e)
+                else if (UInt64.MaxValue - value * 10 < (ulong)(nextByte - '0')) // overflow
                 {
                     value = 0;
                     bytesConsumed = 0;
                     return false;
                 }
+                ulong candidate = checked(value * 10 + nextByte - '0');
```

Comment:
I think the assertion is fine btw, you could even consider adding it back to the other type variants.